File: Assets/Game/Lua/game/scene/actor/actor_trigger_base.lua
ActorTriggerBase = ActorTriggerBase or BaseClass()

function ActorTriggerBase:__init()
	self.delay = 0
	self.transform = nil
	self.enabled = true
	self.target = nil
	self.delay_timer_quest = {}
	self.delay_timer_num = 0
end

function ActorTriggerBase:__delete()
	for _, v in pairs(self.delay_timer_quest) do
		GlobalTimerQuest:CancelQuest(v)
	end
	self.delay_timer_quest = {}
	self.delay_timer_num = 0
end

-- get/set
function ActorTriggerBase:Enalbed(value)
	if value == nil then
		return self.enabled
	end
	self.enabled = value
end

function ActorTriggerBase:OnAnimatorEvent(param, stateInfo, source_obj, target_obj, anim_name)
	if self.enabled then
		local source = source_obj:GetRoot()
		local target = nil
		if target_obj then
			target = target_obj.GetRoot and target_obj:GetRoot() or target_obj
		end

		if self.delay <= 0 then
			self:OnEventTriggered(source, target, source_obj, target_obj, stateInfo)
		else
			self:DelayTrigger(source, target, source_obj, target_obj, stateInfo)
		end
	end
end

function ActorTriggerBase:OnEventTriggered(source, target, source_obj, target_obj, stateInfo)
	-- root source, root target, obj source_obj, obj target_obj
	-- override
end

function ActorTriggerBase:DelayTrigger(source, target, source_obj, target_obj, stateInfo)
	if self.delay_timer_quest[self.delay_timer_num] then
		GlobalTimerQuest:CancelQuest(self.delay_timer_quest[self.delay_timer_num])
	end
	self.delay_timer_quest[self.delay_timer_num] = GlobalTimerQuest:AddDelayTimer(
		function() 
			if source == nil or IsNil(source.gameObject) or not source.gameObject.activeInHierarchy then
				return
			end
			self:OnEventTriggered(source, target, source_obj, target_obj, stateInfo)
			self.delay_timer_num = self.delay_timer_num + 1
		end, self.delay)
end

File: Assets/Game/Scripts/Actor/ActorTriggerBase.cs
//------------------------------------------------------------------------------
// This file is part of MistLand project in Nirvana.
// Copyright © 2016-2016 Nirvana Technology Co., Ltd.
// All Right Reserved.
//------------------------------------------------------------------------------

using System;
using System.Collections;
using Nirvana;
using UnityEngine;

/// <summary>
/// The base class to receive any animation event.
/// </summary>
[Serializable]
public abstract class ActorTriggerBase
{
    [SerializeField]
    [ActorEvent]
    [Tooltip("The event name for this trigger.")]
    private string eventName;

    [SerializeField]
    [Tooltip("The delay time for this trigger.")]
    private float delay;

    private SignalHandle signalHandle;
    private Transform transform;
    private bool enabled = true;
    private IActorTarget target;

    public bool Enalbed
    {
        get { return this.enabled; }
        set { this.enabled = value; }
    }

    public string EventName { get { return eventName; } }

    public float Delay { get { return delay; } }

    /// <summary>
    /// Gets or sets the actor triggers.
    /// </summary>
    protected ActorTriggers ActorTriggers { get; private set; }

    /// <summary>
    /// Gets the actor object.
    /// </summary>
    protected GameObject Actor
    {
        get { return this.transform.gameObject; }
    }

    /// <summary>
    /// Initialize this event trigger.
    /// </summary>
    public virtual void Init(
        ActorTriggers actorTriggers, 
        AnimatorEventDispatcher dispatcher, 
        IActorTarget target)
    {
        this.ActorTriggers = actorTriggers;
        this.transform = dispatcher.transform;
        this.target = target;

        this.signalHandle = dispatcher.ListenEvent(
            this.eventName,
            this.OnAnimatorEvent);
    }

    /// <summary>
    /// Destroy this event trigger.
    /// </summary>
    public virtual void Destroy()
    {
        if (this.signalHandle != null)
        {
            this.signalHandle.Dispose();
            this.signalHandle = null;
        }
    }

    /// <summary>
    /// Update each frame.
    /// </summary>
    public virtual void Update()
    {
        if (this.enabled)
        {
            this.UpdateTrigger();
        }
    }

    /// <summary>
    /// Update each frame.
    /// </summary>
    public virtual void UpdateTrigger()
    {
    }

    /// <summary>
    /// Receive the event triggered.
    /// </summary>
    protected abstract void OnEventTriggered(
        Transform source, Transform target, AnimatorStateInfo stateInfo);

    private void OnAnimatorEvent(string param, AnimatorStateInfo stateInfo)
    {
        if (this.enabled)
        {
            if (this.delay <= 0.0f)
            {
                this.OnEventTriggered(
                    this.transform, this.target.Target, stateInfo);
            }
            else
            {
                Scheduler.RunCoroutine(this.DelayTrigger(
                    this.transform, this.target.Target, stateInfo));
            }
        }
    }

    private IEnumerator DelayTrigger(
        Transform source, Transform target, AnimatorStateInfo stateInfo)
    {
        yield return new WaitForSeconds(this.delay);

        if (this == null || 
            this.transform == null || 
            this.transform.gameObject == null || 
            !this.transform.gameObject.activeInHierarchy)
        {
            yield break;
        }

        this.OnEventTriggered(source, target, stateInfo);
    }
}
